### DISCUSSION TITLE
Button is not working after refresh or page switch

### DISCUSSION THREAD

**Discussion by Swasthik-Jain** [zauberzeug/nicegui#3140] [Q&A]
### Question

I wanted to achieve this 

if toggled, do a first load
if toggled again, hide 
if toggled again, show the previously loaded kpi, run a refresh in background --> should be much faster and updates values if changes have occurred

I have a main page called search.py from there I'm calling 
```py
async def create_search_panel() -> None:
    with ui.row().classes('w-full'):
        ui.label("Search").classes("text-bold text-xl mt-2")
        ui.button('View Insights', icon='dvr', on_click=toggle_kpi_panel)
 ```
 
I have another page where I defined kpi_panel in kpi.py 
 
```py
 kpi = None
async def toggle_kpi_panel():
    global kpi
    if kpi is None:
        kpi = await kpi_panel()
    else:
        kpi.toggle()

def handle_tab_change():
    global kpi
    if kpi is not None and hasattr(kpi, 'hide'):
        kpi.hide()

async def kpi_panel():
    global kpi
    if kpi is None:
        with ui.right_drawer() as kpi:
            with ui.row().classes(
                    "flex items-center flex-row p-4 left-0 fixed top-[-88.8px] w-[302px] h-24 bg-gray-100"):
                ui.label("Insights ").classes("text-xl font-bold absolute")
                close_button = ui.icon('close').classes('text-2xl cursor-pointer absolute right-8').tooltip(
                    'Close insights')

            async def close_kpi():
                kpi.hide()

            close_button.on('click', close_kpi)
            spinner = Spinner().classes('flex justify-center items-center w-full h-full')
            # Displaying other contents
           spinner.set_visibility(False)
    return kpi
```

When the page first loads toggling function works fine[show/hide] and even the close button works fine
  if I switch to any other page or refresh the page, I'm not able to trigger the kpi_panel again [ button will not work]
  I need your guidance on this matter how to achieve above statements stated in the beginning.

**Comment by falkoschindler** [comment]
Hi @Swasthik-Jain,

Please try to provide a minimal executable example. This allows us and the community to help more efficiently. Thanks!
[Why is it important to provide a minimal reproducible example?](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example)

**Reply by Swasthik-Jain** [reply]
Hi @falkoschindler ,

Here is the sample code 
As per my need I wanted the toggle to work as follows
if toggled, do a first load
if toggled again, hide
if toggled again, show the previously loaded kpi, run a refresh in background --> should be much faster and updates values if changes have occurred

Toggling function works fine when we load the UI first-time 
[ After clicking the button view insight, it loads kpi panel and we can toggle [show /hide]
But the problem occurs when we refresh or switch to other pages 
Regarding refresh I need you are guidance how can achieve to load new data if available

```py
from nicegui import ui
from pages.search import render_search_page
from pages.process import render_process_page
from pages.line import render_line_page

@ui.page('/')
def index():
    ui.label('Welcome to the Home Page')
    ui.link('Go to Search Page', '/search')
    ui.link('Go to Process Page', '/process')
    ui.link('Go to Line Page', '/line')

@ui.page('/search')
async def search():
    await render_search_page()

@ui.page('/process')
def process():
    render_process_page()

@ui.page('/line')
def line():
    render_line_page()

if __name__ in {"__main__", "__mp_main__"}:
    ui.run()
```

For example, I have 3 pages under pages folder

Search.py
```py
from nicegui import ui
from theme.kpi import handle_tab_change, toggle_kpi_panel

_INITIAL_LOAD_TIMEOUT = 30.0

def create_api_panel():
    ui.label("Api panel")

def create_explore_panel():
    ui.label('Explore panel')

async def create_search_panel():
    with ui.row().classes('w-full'):
        ui.label("Search").classes("text-bold text-xl mt-2")
        ui.button('View Insights', icon='dvr', on_click=toggle_kpi_panel).classes('ml-auto')

async def render_search_page():
    ui.label('This is the Search Page')

    with ui.tabs(on_change=handle_tab_change).classes('w-full lis-tab-container') as tabs:
        tabs_styling = 'text-black lis-tab'
        ui.tab('Search', icon='search').classes(tabs_styling)
        ui.tab('Explore', icon='explore').classes(tabs_styling)
        ui.tab('API', icon='api').classes(tabs_styling)

    with ui.tab_panels(tabs, value='Search').classes('w-full'):
        with ui.tab_panel('Search').classes('w-full'):
            with ui.element('div').classes('w-full').style('height: calc(100vh - 255px);'):
                await create_search_panel()

        with ui.tab_panel('Explore'):
            create_explore_panel()

        with ui.tab_panel('API'):
            with ui.element('div').classes('w-full').style('height: calc(100vh - 255px);'):
                create_api_panel()
```

process.py
```py
from nicegui import ui
def render_process_page():
    ui.label('This is the Process Page')
```

line.py
```py
from nicegui import ui
def render_line_page():
    ui.label('This is the Line Page')
```

Under theme folder I have kpi component
```py
from nicegui import ui

kpi = None

async def toggle_kpi_panel():
    global kpi
    if kpi is None:
        kpi = await kpi_panel()
    else:
        kpi.toggle()

def handle_tab_change():
    global kpi
    if kpi is not None and hasattr(kpi, 'hide'):
        kpi.hide()

async def kpi_panel():
    global kpi
    if kpi is None:
        with ui.right_drawer().classes('text-gray-700') as kpi:
            with ui.element('div').classes(
                    "flex items-center p-4 fixed top-[-88.8px] w-[302px] h-24 bg-gray-100 left-0"):
                ui.label("Insights ").classes("text-xl font-bold")
                close_button = ui.icon('close').classes('text-2xl cursor-pointer absolute right-8').tooltip(
                    'Close insights')

            async def close_kpi():
                kpi.hide()

            close_button.on('click', close_kpi)
            custom_styles = """
            .q-drawer.q-drawer--right.q-drawer--standard.fixed {
                top: 218px !important;
                background: rgb(243, 244, 246) !important;
            }
            """
            ui.add_head_html(f'<style>{custom_styles}</style>')

            for i in range(1, 11):
                with ui.card().classes('w-full'):
                    ui.label(f'card{i}')
            spinner = ui.spinner().classes('flex justify-center items-center w-full h-full')
            spinner.set_visibility(False)

    return kpi
```

**Reply by falkoschindler** [reply]
> When creating an MRE, make sure to remove any code that is not directly related to the issue, while keeping the example complete and runnable. This includes imports, functions, and data. Keep the example as simple as possible while still demonstrating the problem.

**Reply by Swasthik-Jain** [reply]
Hi @falkoschindler,
Minimal executable example.
```py 
from nicegui import ui
from kpi import toggle_kpi_panel

async def create_search_panel():
    with ui.row().classes('w-full'):
        ui.label("Search").classes("text-bold text-xl mt-2")
        ui.button('View Insights', icon='dvr', on_click=toggle_kpi_panel).classes('ml-auto')
@ui.page('/')
async def search():
    ui.label('This is the Search Page')
    with ui.tab_panel('Search').classes('w-full'):
        with ui.element('div').classes('w-full').style('height: calc(100vh - 255px);'):
            await create_search_panel()

if __name__ in {"__main__", "__mp_main__"}:
    ui.run()
```

```py
from nicegui import ui

kpi = None
async def toggle_kpi_panel():
    global kpi
    if kpi is None:
        kpi = await kpi_panel()
    else:
        kpi.toggle()

async def kpi_panel():
    global kpi
    if kpi is None:
        with ui.right_drawer().classes('text-gray-700') as kpi:
            with ui.element('div').classes(
                    "flex items-center p-4 fixed top-[-88.8px] w-[302px] h-24 bg-gray-100 left-0"):
                ui.label("Insights ").classes("text-xl font-bold")
                close_button = ui.icon('close').classes('text-2xl cursor-pointer absolute right-8').tooltip(
                    'Close insights')
            async def close_kpi():
                kpi.hide()
            close_button.on('click', close_kpi)
            custom_styles = """
            .q-drawer.q-drawer--right.q-drawer--standard.fixed {
                top: 154px !important;
                background: rgb(243, 244, 246) !important;
            }
            """
            ui.add_head_html(f'<style>{custom_styles}</style>')

            for i in range(1, 11):
                with ui.card().classes('w-full'):
                    ui.label(f'card{i}')
            spinner = ui.spinner().classes('flex justify-center items-center w-full h-full')
            spinner.set_visibility(False)

    return kpi
```

**Reply by falkoschindler** [reply]
> When the page first loads toggling function works fine[show/hide] and even the close button works fine

The close button doesn't work for me. But it can be fixed, e.g., like this:
```py
close_button.on('click', kpi.hide)
```

> But the problem occurs when we refresh

The "View Insights" button still works after a refresh.

All in all, your state management is a bit strange. There's one major bug I noticed: You're assigning `kpi` in this line:
```py
with ui.right_drawer().classes('text-gray-700') as kpi:
```
But then you assign it again to the return value of `kpi_panel()`:
```py
kpi = await kpi_panel()
```
This will set `kpi` back to `None`, because `kpi_panel` doesn't return anything. Removing the later assignment also solves the `kpi.hide` problem mentioned above.

**Reply by Swasthik-Jain** [reply]
Hi @falkoschindler, could you please reproduce the code with your changes to help me understand it better?

**Reply by falkoschindler** [reply]
Find `kpi = await kpi_panel()`, replace it with `await kpi_panel()`.

**Reply by Swasthik-Jain** [reply]
Sorry to disturb again "View Insights" button isn't functioning properly for me after refreshing the page or switching to another page.

**Reply by falkoschindler** [reply]
Ok, I was wrong. `kpi_panel` _is_ returning `kpi`, so my fix doesn't work.

I think the real problem is that you have only one global `kpi` variable holding a `ui.right_drawer` object. But when serving multiple clients via `@ui.page('/')`, the second visitor doesn't get a new drawer created. I'm not sure why I missed that the first time. Probably because your example was still quite complex.

Let me show you a _minimal_ reproduction of your problem:
```py
kpi = None

def toggle_drawer():
    global kpi
    if kpi is None:
        with ui.right_drawer() as kpi:
            ui.label('Drawer')
    else:
        kpi.toggle()

@ui.page('/')
def page():
    ui.button('Show drawer', on_click=toggle_drawer)
```

You can solve it by nesting the `kpi` object in the page function:

```py
@ui.page('/')
def page():
    kpi = None

    def toggle_drawer():
        nonlocal kpi
        if kpi is None:
            with ui.right_drawer() as kpi:
                ui.label('Drawer')
        else:
            kpi.toggle()

    ui.button('Show drawer', on_click=toggle_drawer)
```

Much better would be to create the drawer immediately and to toggle it on button click:

```py
@ui.page('/')
def page():
    with ui.right_drawer() as kpi:
        ui.label('Drawer')

    ui.button('Show drawer', on_click=kpi.toggle)
```

**Comment by Swasthik-Jain** [comment]
Thank you for your response. Yes, your approach works well, but the issue is that it reloads whenever the page is refreshed or when switching pages.
I have previously done it in the same way.
loading the data when the page is loading then using the button to toggle
```py
from nicegui import ui

kpi = ui.right_drawer()
async def toggle_drawer():
    kpi.toggle()

async def kpi_panel():
    global kpi
    with ui.right_drawer(value=False) as kpi:
        ui.label('Drawer')
    return kpi

@ui.page('/')
async def page():
    ui.button('Show drawer', on_click=toggle_drawer)
    await kpi_panel()

ui.run()
```
This setup functions effectively, similar to your approach, as we are simply displaying labels, and it works well.
For instance, if we have 10 cards that need to fetch data via a query, the page will only load after the data has been retrieved for the cards.
 To address this, I am considering implementing a button that, when activated, will reveal the content, thus allowing the page to load more swiftly. Once the data is available, I want to avoid reloading it, as it is time-consuming. My proposed approach is as follows:

- On toggle, perform the initial load.
- On re-toggle, hide the content.
- On toggle once more, display the previously loaded KPIs and execute a background refresh, which should be quicker and update the values if any changes have occurred.

**Reply by falkoschindler** [reply]
Ok, I see. I strongly recommend separating UI from your data model. If you want to fetch KPIs only once, do it and store it in a global `kpi` object (e.g. a dictionary or other data structure). The UI can be created independently, _presenting_ the data stored in `kpi`. If there is no data yet, it can trigger and await the load and render it afterwards.
